File: codetraverse/ast_diff/haskelldiff.py
import json
from tree_sitter import Language, Parser, Node
import tree_sitter_haskell
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff

class HaskellFileDiff(BaseFileDiff):
    """Analyzes and compares two Haskell ASTs for semantic differences."""
    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node) -> str:
        """Finds the name of a declaration. For instances, it extracts the instance head."""
        # For instances, the "name" is the instance head (e.g., "Eq (Maybe a)")
        if node.type == 'instance':
            # The instance head is typically the part between 'instance' and 'where'
            instance_head_nodes = []
            for child in node.children:
                if child.type == 'where':
                    break
                if child.type != 'instance':
                    instance_head_nodes.append(child.text.decode(errors="ignore"))
            return " ".join(instance_head_nodes).strip()
        
        # For class declarations, look specifically for the class name
        if node.type == 'class':
            for child in node.children:
                if child.type == 'name':  # Class names are in 'name' nodes
                    return child.text.decode(errors="ignore")
            # Fallback: get first identifier after 'class'
            for i, child in enumerate(node.children):
                if child.type == 'class' and i + 1 < len(node.children):
                    next_child = node.children[i + 1]
                    if next_child.type in ['name', 'constructor', 'variable']:
                        return next_child.text.decode(errors="ignore")

        # For other types, find the first variable or constructor
        queue = list(node.children)
        while queue:
            current = queue.pop(0)
            if current.type in ["variable", "constructor"]:
                return current.text.decode(errors="ignore")
            if current.is_named:
                queue.extend(current.children)
        return None

    def extract_components(self, root: Node):
        """Extracts all top-level Haskell declarations from an AST."""
        functions, data_types, type_classes, instances, imports, template_haskell = {}, {}, {}, {}, {}, {}
        
        node_type_map = {
            "function": functions,
            "signature": functions,
            "bind": functions,  # Function implementations
            "data_type": data_types,
            "class": type_classes,
            "instance": instances,  # Track instances
            "import": imports,
            "top_splice": template_haskell,  # Template Haskell splices
        }
        
        declarations = []
        if root.type == 'haskell':
            # Look for declarations node
            declarations_node = next((c for c in root.children if c.type == 'declarations'), None)
            if declarations_node:
                declarations = declarations_node.children
            
            # Look for imports node separately
            imports_node = next((c for c in root.children if c.type == 'imports'), None)
            if imports_node:
                for import_child in imports_node.children:
                    if import_child.type == "import":
                        # For imports, use the full import text as the name
                        name = import_child.text.decode(errors="ignore").strip()
                        imports[name] = (import_child, import_child.text.decode(errors="ignore"), import_child.start_point, import_child.end_point)
        
        # Process declarations (functions, data types, etc.)
        # We need to combine type signatures with function implementations
        for child in declarations:
            if child.type in node_type_map and child.type != "import":
                name = self.get_decl_name(child)
                if name:
                    target_dict = node_type_map[child.type]
                    
                    if name not in target_dict:
                        # First time seeing this name
                        target_dict[name] = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)
                    else:
                        # We've seen this name before - combine signature and implementation
                        existing_node, existing_text, start, end = target_dict[name]
                        new_text = child.text.decode(errors="ignore")
                        
                        # Determine which is signature and which is implementation
                        if child.type == "signature":
                            # Current is signature, existing might be function
                            combined_text = new_text + "\n" + existing_text
                            new_start = child.start_point
                            new_end = end
                        elif existing_node.type == "signature":
                            # Existing is signature, current is function  
                            combined_text = existing_text + "\n" + new_text
                            new_start = start
                            new_end = child.end_point
                        else:
                            # Both are same type, just concatenate
                            combined_text = existing_text + "\n" + new_text
                            new_start = start
                            new_end = child.end_point
                        
                        target_dict[name] = (child, combined_text, new_start, new_end)
        return functions, data_types, type_classes, instances, imports, template_haskell

    def diff_components(self, before_map: dict, after_map: dict):
        """Compares two dictionaries of components and returns the diff."""
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())
        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common_names = before_names & after_names
        
        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]
        modified = []
        for name in sorted(common_names):
            _, old_body, _, _ = before_map[name]
            _, new_body, old_start, old_end = after_map[name]
            if old_body.strip() != new_body.strip():
                 modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end}))
        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast: Node, new_file_ast: Node) -> DetailedChanges:
        """The main method to compare two Haskell files."""
        old_funcs, old_data, old_classes, old_instances, old_imports, old_th = self.extract_components(old_file_ast.root_node)
        new_funcs, new_data, new_classes, new_instances, new_imports, new_th = self.extract_components(new_file_ast.root_node)

        category_map = {
            "functions": (old_funcs, new_funcs),
            "dataTypes": (old_data, new_data),
            "typeClasses": (old_classes, new_classes),
            "instances": (old_instances, new_instances),
            "imports": (old_imports, new_imports),
            "templateHaskell": (old_th, new_th),
        }

        for category, (old_map, new_map) in category_map.items():
            diff = self.diff_components(old_map, new_map)
            for change_type in ["added", "deleted", "modified"]:
                for item in diff[change_type]:
                    self.changes.add_change(category, change_type, item)

        return self.changes
        
    def process_single_file(self, file_ast: Node, mode="deleted") -> DetailedChanges:
        funcs, data, classes, instances, imports, template_haskell = self.extract_components(file_ast.root_node)

        category_map = {
            "functions": funcs,
            "dataTypes": data,
            "typeClasses": classes,
            "instances": instances,
            "imports": imports,
            "templateHaskell": template_haskell,
        }

        for category, component_map in category_map.items():
            for name, data_tuple in component_map.items():
                # data_tuple is (node, text, start_point, end_point)
                item = (name, data_tuple[1], {"start": data_tuple[2], "end": data_tuple[3]})
                self.changes.add_change(category, mode, item)
        
        return self.changes

File: codetraverse/ast_diff/haskelldiff.ts
import { SyntaxNode } from 'tree-sitter';
import { DetailedChanges } from './Detailedchanges';
import { BaseFileDiff } from './basefilediff';

export class HaskellFileDiff extends BaseFileDiff {
    private parser: any;

    constructor(moduleName: string = "", parser: any) {
        super(moduleName);
        this.changes = new DetailedChanges(moduleName);
        this.parser = parser;
    }

    getDeclName(node: SyntaxNode): string | null {
        if (node.type === 'instance') {
            const instanceHeadNodes: string[] = [];
            for (const child of node.children) {
                if (child.type === 'where') {
                    break;
                }
                if (child.type !== 'instance') {
                    instanceHeadNodes.push(child.text);
                }
            }
            return instanceHeadNodes.join(" ").trim();
        }

        if (node.type === 'class') {
            for (const child of node.children) {
                if (child.type === 'name') {
                    return child.text;
                }
            }
            for (let i = 0; i < node.children.length; i++) {
                const child = node.children[i];
                if (child.type === 'class' && i + 1 < node.children.length) {
                    const nextChild = node.children[i + 1];
                    if (['name', 'constructor', 'variable'].includes(nextChild.type)) {
                        return nextChild.text;
                    }
                }
            }
        }

        const queue = [...node.children];
        while (queue.length > 0) {
            const current = queue.shift()!;
            if (['variable', 'constructor'].includes(current.type)) {
                return current.text;
            }
            if (current.isNamed) {
                queue.push(...current.children);
            }
        }
        return null;
    }

    extractComponents(root: SyntaxNode): any[] {
        const functions: any = {};
        const data_types: any = {};
        const type_classes: any = {};
        const instances: any = {};
        const imports: any = {};
        const template_haskell: any = {};

        const nodeTypeMap: Record<string, any> = {
            "function": functions,
            "signature": functions,
            "bind": functions,
            "data_type": data_types,
            "class": type_classes,
            "instance": instances,
            "import": imports,
            "top_splice": template_haskell,
        };

        let declarations: SyntaxNode[] = [];
        if (root.type === 'haskell') {
            const declarationsNode = root.children.find(c => c.type === 'declarations');
            if (declarationsNode) {
                declarations = declarationsNode.children;
            }

            const importsNode = root.children.find(c => c.type === 'imports');
            if (importsNode) {
                for (const importChild of importsNode.children) {
                    if (importChild.type === "import") {
                        const name = importChild.text.trim();
                        imports[name] = [importChild, importChild.text, importChild.startPosition, importChild.endPosition];
                    }
                }
            }
        }

        for (const child of declarations) {
            if (child.type in nodeTypeMap && child.type !== "import") {
                const name = this.getDeclName(child);
                if (name) {
                    const targetDict = nodeTypeMap[child.type];
                    if (!targetDict[name]) {
                        targetDict[name] = [child, child.text, child.startPosition, child.endPosition];
                    } else {
                        const [existingNode, existingText, start, end] = targetDict[name];
                        const newText = child.text;
                        let combinedText: string;
                        let newStart: any;
                        let newEnd: any;

                        if (child.type === "signature") {
                            combinedText = newText + "\n" + existingText;
                            newStart = child.startPosition;
                            newEnd = end;
                        } else if (existingNode.type === "signature") {
                            combinedText = existingText + "\n" + newText;
                            newStart = start;
                            newEnd = child.endPosition;
                        } else {
                            combinedText = existingText + "\n" + newText;
                            newStart = start;
                            newEnd = child.endPosition;
                        }
                        targetDict[name] = [child, combinedText, newStart, newEnd];
                    }
                }
            }
        }
        return [functions, data_types, type_classes, instances, imports, template_haskell];
    }

    diffComponents(beforeMap: any, afterMap: any): any {
        const beforeNames = new Set(Object.keys(beforeMap));
        const afterNames = new Set(Object.keys(afterMap));

        const addedNames = Array.from(afterNames).filter(name => !beforeNames.has(name));
        const deletedNames = Array.from(beforeNames).filter(name => !afterNames.has(name));
        const commonNames = Array.from(beforeNames).filter(name => afterNames.has(name));

        const added = addedNames.sort().map(n => [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]);
        const deleted = deletedNames.sort().map(n => [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]);

        const modified = [];
        for (const name of commonNames.sort()) {
            const oldBody = beforeMap[name][1];
            const newBody = afterMap[name][1];
            if (oldBody.trim() !== newBody.trim()) {
                modified.push([name, oldBody, newBody, { old_start: afterMap[name][2], old_end: afterMap[name][3] }]);
            }
        }

        return { added, deleted, modified };
    }

    compareTwoFiles(oldFileAst: SyntaxNode, newFileAst: SyntaxNode): DetailedChanges {
        const [oldFuncs, oldData, oldClasses, oldInstances, oldImports, oldTH] = this.extractComponents(oldFileAst);
        const [newFuncs, newData, newClasses, newInstances, newImports, newTH] = this.extractComponents(newFileAst);

        const categoryMap = {
            "functions": [oldFuncs, newFuncs],
            "dataTypes": [oldData, newData],
            "typeClasses": [oldClasses, newClasses],
            "instances": [oldInstances, newInstances],
            "imports": [oldImports, newImports],
            "templateHaskell": [oldTH, newTH],
        };

        for (const category in categoryMap) {
            const [oldMap, newMap] = categoryMap[category as keyof typeof categoryMap];
            const diff = this.diffComponents(oldMap, newMap);
            for (const changeType in diff) {
                for (const item of diff[changeType]) {
                    this.changes.add_change(category, changeType, item);
                }
            }
        }

        return this.changes;
    }

    processSingleFile(fileAst: SyntaxNode, mode: string = "deleted"): DetailedChanges {
        const [funcs, data, classes, instances, imports, template_haskell] = this.extractComponents(fileAst);

        const categoryMap = {
            "functions": funcs,
            "dataTypes": data,
            "typeClasses": classes,
            "instances": instances,
            "imports": imports,
            "templateHaskell": template_haskell,
        };

        for (const category in categoryMap) {
            const componentMap = categoryMap[category as keyof typeof categoryMap];
            for (const name in componentMap) {
                const dataTuple = componentMap[name];
                const item = [name, dataTuple[1], { start: dataTuple[2], end: dataTuple[3] }];
                this.changes.add_change(category, mode, item);
            }
        }

        return this.changes;
    }
}
